---
title: "WTS - bcbio VS dragen results"
author: "Sehrish Kanwal (UMCCR)"
date: "`r format(Sys.time(), '%a %Y-%b-%d')`"
output:
  html_document:
    keep_md: yes
    css: style.css
    theme: readable
    toc: true
    toc_float: true
    code_folding: hide
  rmdformats::material:
    highlight: kate
params:
  run_1: '/Users/kanwals/UMCCR/data/projects/WTS_comparison/SBJ02407/dragen/'
  run_2: '/Users/kanwals/UMCCR/data/projects/WTS_comparison/SBJ02407/bcbio/'
---

# Description

<style>
div.blue { background-color:#e6f0ff; border-radius: 5px; padding: 20px;}
</style>
<div class = "blue">

**SAMPLE: SBJ02407**

For each WTS workflow output type (abundance and fusions for now), the document currently reports:

- File names in the input (parms) directories
- The number and comparison of calls
- The key difference between both outputs (i.e. bcbio VS Dragen) in a tabular format

</div>


```{r setup, include=FALSE}
knitr::opts_chunk$set(collapse = TRUE) # global setup
```

```{r define_functions, comment=NA, message=FALSE, warning=FALSE}
### Define functions.
###list files in a directory
lf <- function(...) {
  data.frame(fname = list.files(...)) %>% 
    knitr::kable(row.names = TRUE)
}

###generate scatter and violin plot for tpm values across both runs
scatterPlot <- function(icombined_tpm, main_title, plot_mode = "static") {
  plotly::plot_ly(icombined_tpm, x = ~log2(tpm_bcbio+0.5), y = ~log2(tpm_dragen+0.5), type = 'scattergl', 
                       text = icombined_tpm$target_id, width = 00, height = 400,
                               marker = list(size = 10,
                       color = 'rgba(255, 182, 193, .5)',
                       #opacity = 0.5, https://plot.ly/~alex/455/four-ways-to-change-opacity-of-scatter-markers.embed
                       line = list(color = 'rgba(152, 0, 0, .8)',
                                   width = 2))) %>%
    plotly::layout(title = main_title)
}

###VIOLIN plot https://plot.ly/r/violin/
violinPlot <- function(icombined_tpm, main_title, plot_mode = "static") {
  #overall data distribution
  p1 <- plotly::plot_ly(icombined_tpm, type = 'violin') %>%
    add_trace(y = ~log2(tpm_bcbio+1), box = list(visible = T),  meanline = list(visible = F),
                       x = 'bcbio', legendgroup = 'run1', text = icombined_tpm$target_id, name = 'bcbio', color = 'orange') %>%
    add_trace(y = ~log2(tpm_dragen+1), box = list(visible = T),  meanline = list(visible = F),
                       x = 'dragen', legendgroup = 'run2', text = icombined_tpm$target_id, name = 'dragen', color = 'green') %>%
    layout(yaxis = list(title = "log2 tpm", zeroline = F))
  
  #zoom into data to highlight data spread
  p2 <- plotly::plot_ly(icombined_tpm, type = 'violin') %>%
    add_trace(y = c(~log2(quantile(tpm_bcbio+1)[1]), ~log2(quantile(tpm_bcbio+1)[4])), box = list(visible = T),  meanline = list(visible = F),
                       x = 'bcbio', legendgroup = 'run1', text = icombined_tpm$target_id, name = 'bcbio', color = 'orange') %>%
    add_trace(y = c(~log2(quantile(tpm_dragen+1)[1]), ~log2(quantile(tpm_dragen+1)[4])), box = list(visible = T),  meanline = list(visible = F),
              x = 'dragen', legendgroup = 'run2', text = icombined_tpm$target_id, name = 'dragen', color = 'green') %>%
    layout(yaxis = list(title = "log2 tpm", zeroline = F))
  
  #combine both box plots
  subplot(p1, p2, shareX = FALSE, shareY = FALSE, titleY = TRUE, margin = 0.06) %>%
    hide_legend()
}

```


```{r, message=FALSE, warning=FALSE}
### Required packages.
library(tools)
library(here)
library(DT)
library(dplyr)
library(ggplot2)
library(plotly)
library(gridExtra)
library(tximport)
library(package="EnsDb.Hsapiens.v86", character.only = TRUE)
```

***

# List files

This section lists the file names for outputs in the run directories being compared in this report.  

```{r read_data, message=FALSE, warning=FALSE}
###Read input. 
###The script expects two input directories, each containing 'final' results as produced by bcbio. The results in these two input directories will be compared.
run_1 <-file.path(params$run_1)
#list files in the first run directory
list.files(run_1)

run_2 <- file.path(params$run_2)
#list files in the second run directory
list.files(run_2)

```

***

# Counts

This section summarizes two key factors:

- a summary of the total number of counts in the compared files and 
- a table outlining difference between both runs.

```{r input-counts, message=FALSE, warning=FALSE}
###Start with the simple line count in the files i.e. the number of entries in the output from both runs is same.
#compare the count files first (from both runs)
counts.dragen <- read.table(here("../../data/projects/WTS_comparison/SBJ02407/dragen/PRJ221346.quant.sf"), sep="\t", as.is=TRUE, header=TRUE)
counts.dragen.gene <- read.table(here("../../data/projects/WTS_comparison/SBJ02407/dragen/PRJ221346.quant.genes.sf"), sep="\t", as.is=TRUE, header=TRUE)
counts.bcbio <-  read.table(here("../../data/projects/WTS_comparison/SBJ02407/bcbio/abundance.tsv"), sep="\t", as.is=TRUE, header=TRUE)

#read in the tximport data
dir <- here("../../data/projects/WTS_comparison")
tx2ensembl <- read.csv(file.path(dir, "tx2gene.ensembl.v95.csv"), header = FALSE)
colnames(tx2ensembl) <- c("tx_name", "gene_id")

#read in UMCCR cancer gene list
umccr.cancer.genes <- read.table(file = file.path(dir, "umccr_cancer_genes.latest.tsv"), sep = '\t', header = TRUE)
```

### Total counts

```{r total-counts, message=FALSE, warning=FALSE}
###calculating number of lines in count files from both runs
lines.count <- data.frame(c("dragen", "bcbio"), c(nrow(counts.dragen), nrow(counts.bcbio)))
colnames(lines.count) <- c("file", "lines_count")
datatable(lines.count, rownames = TRUE, options = list(sDom  = '<"top">lrt<"bottom">ip'))
```

### Diff counts

```{r diff-counts, message=FALSE, warning=FALSE, echo =FALSE}
###find the differences between count files from both runs
#dragen names columns differently. Changing these to reflect bcbio's output format, making the comparison's easy.
colnames(counts.dragen) <-  c("target_id", "length", "eff_length", "tpm", "est_counts")
counts.dragen <- counts.dragen[, c(1, 2, 3, 5, 4)]

#extract gene level counts using tximport
txi.bcbio <- tximport(here('../../data/projects/WTS_comparison/SBJ02407/bcbio/abundance.tsv'), type = "kallisto", tx2gene = tx2ensembl)
txi.dragen <- tximport(here('../../data/projects/WTS_comparison/SBJ02407/dragen/PRJ221346.quant.sf'), type = "salmon", tx2gene = tx2ensembl)

#extract counts to prepare dataframe
bcbio.counts <- as.data.frame(txi.bcbio$counts) %>%
  tibble::rownames_to_column("gene") %>%
  dplyr::rename(count = V1)

dragen.counts <- as.data.frame(txi.dragen$counts) %>%
  tibble::rownames_to_column("gene") %>%
  dplyr::rename(count = V1)

dragen.counts.native <- as.data.frame(counts.dragen.gene) %>%
  dplyr::select(Name, NumReads) %>%
  dplyr::rename(gene = Name,
                count= NumReads)

#merge the calls from bcbio and dragen - also add a column for recording percentage difference
bcbio.dragen.counts <- inner_join(bcbio.counts, dragen.counts, by = "gene") %>%
  dplyr::mutate(percentageDiff = ifelse(count.x > count.y, (count.x - count.y)/count.x *100, (count.y - count.x)/count.y *100)) %>%
  dplyr::mutate_if(is.numeric, round,2)
colnames(bcbio.dragen.counts) <- c("gene", "bcbio", "dragen", "percentageDiff")

#merge the calls from bcbio and dragen native gene count estimation - also add a column for recording percentage difference
bcbio.dragennative.counts <- inner_join(bcbio.counts, dragen.counts.native, by = "gene") %>%
  dplyr::mutate(percentageDiff = ifelse(count.x > count.y, (count.x - count.y)/count.x *100, (count.y - count.x)/count.y *100)) %>%
  dplyr::mutate_if(is.numeric, round,2)
colnames(bcbio.dragennative.counts) <- c("gene", "bcbio", "dragen_native", "percentageDiff") 

#merge the calls from dragen-salmon-to-gene using txi and dragen native gene count estimation - also add a column for recording percentage difference
dragen.dragennative.counts <- inner_join(dragen.counts, dragen.counts.native, by = "gene") %>%
  dplyr::mutate(percentageDiff = ifelse(count.x > count.y, (count.x - count.y)/count.x *100, (count.y - count.x)/count.y *100)) %>%
  dplyr::mutate_if(is.numeric, round,2)
colnames(dragen.dragennative.counts) <- c("gene", "dragen_txi", "dragen_native", "percentageDiff") 

#get genes annotation from ensembl IDs 
edb <- EnsDb.Hsapiens.v86
#get keytypes for gene SYMBOL
keys <- keys(edb, keytype="GENEID")
#get genes genomic coordiantes
gene_info <- ensembldb::select(edb, keys=keys, columns=c("GENEID", "GENENAME"), keytype="GENEID")
colnames(gene_info) <- c("gene", "geneName")
#remove rows with duplicated ENSEMBL IDs - if any
gene_info = gene_info[!duplicated(gene_info$gene),]
gene_info = gene_info[!duplicated(gene_info$geneName),]

#annotate the counts dataframe to include gene names
bcbio.dragen.gene.counts <- inner_join(bcbio.dragen.counts, gene_info, by = "gene") %>%
  dplyr::arrange(-percentageDiff) %>%
  dplyr::select(geneName, everything()) %>%
  #remove part of genenames after .
  dplyr::mutate(geneName=sapply(strsplit(geneName, split = '.', fixed = TRUE), function(x) (x[1])))

#annotate the dragen native counts dataframe to include gene names
bcbio.dragennative.gene.counts <- inner_join(bcbio.dragennative.counts, gene_info, by = "gene") %>%
  dplyr::arrange(-percentageDiff) %>%
  dplyr::select(geneName, everything()) %>%
  #remove part of genenames after .
  dplyr::mutate(geneName=sapply(strsplit(geneName, split = '.', fixed = TRUE), function(x) (x[1])))

#annotate the dragen-dragennative counts dataframe to include gene names
dragen.dragennative.gene.counts <- inner_join(dragen.dragennative.counts, gene_info, by = "gene") %>%
  dplyr::arrange(-percentageDiff) %>%
  dplyr::select(geneName, everything()) %>%
  #remove part of genenames after .
  dplyr::mutate(geneName=sapply(strsplit(geneName, split = '.', fixed = TRUE), function(x) (x[1])))

##merge the raw estimated calls from bcbio and dragen - also add a column for recording percentage difference
bcbio.dragen.transcript.counts <- inner_join(counts.bcbio, counts.dragen, by = "target_id") %>%
  dplyr::mutate(percentageDiff = ifelse(est_counts.x > est_counts.y, (est_counts.x - est_counts.y)/est_counts.x *100, (est_counts.y - est_counts.x)/est_counts.y *100)) %>%
  dplyr::mutate_if(is.numeric, round,2)
colnames(bcbio.dragen.transcript.counts) <- c("target_id", "length.bcbio", "eff_length.bcbio", "est_counts.bcbio", "tpm.bcbio", "length.dragen", "eff_length.dargen", "est_counts.dragen", "tpm.dragen", "percentage_counts_diff")

#subset the df to only keep values and percentage difference
est.counts <- bcbio.dragen.transcript.counts[, c('target_id', 'est_counts.bcbio', 'est_counts.dragen', 'percentage_counts_diff')]

#Per transcript counts correlation between bcbio-kallisto and dragen
cor.transcript = cor.test(log2(est.counts$est_counts.bcbio+1), log2(est.counts$est_counts.dragen+1), method = 'spearman')
ggplot(est.counts, 
       aes(x=log2(est.counts$est_counts.bcbio+1), y=log2(est.counts$est_counts.dragen+1))) +
         geom_point() +
         geom_smooth(method = "lm") +
         geom_abline(slope=1, intercept = 0, color="red") +
         annotate("text", x=15, y=20, label= paste("spearman cor = ", round(cor.transcript$estimate, 2))) +
         ggtitle("Per transcript counts correlation between bcbio-kallisto and dragen")

#Per gene counts correlation between bcbio-kallisto and dragen
cor.gene = cor.test(log2(bcbio.dragen.gene.counts$bcbio+1), log2(bcbio.dragen.gene.counts$dragen+1), method = 'spearman')
ggplot(bcbio.dragen.gene.counts, 
       aes(x=log2(bcbio.dragen.gene.counts$bcbio+1), y=log2(bcbio.dragen.gene.counts$dragen+1))) +
         geom_point() +
         geom_smooth(method = "lm") +
         geom_abline(slope=1, intercept = 0, color="red") +
         annotate("text", x=15, y=20, label= paste("spearman cor = ", round(cor.gene$estimate, 2))) +
         ggtitle("Per gene counts correlation between bcbio-kallisto and dragen")

#Per gene counts correlation between bcbio-kallisto-to-gene and dragen-native-gene
cor.gene.dragennative = cor.test(log2(bcbio.dragennative.gene.counts$bcbio+1),
                                 log2(bcbio.dragennative.gene.counts$dragen_native+1), method = 'spearman')
ggplot(bcbio.dragennative.gene.counts, 
       aes(x=log2(bcbio.dragennative.gene.counts$bcbio+1), y=log2(bcbio.dragennative.gene.counts$dragen_native+1))) +
         geom_point() +
         geom_smooth(method = "lm") +
         geom_abline(slope=1, intercept = 0, color="red") +
         annotate("text", x=15, y=20, label= paste("spearman cor = ", round(cor.gene.dragennative$estimate, 2))) +
         ggtitle("Per gene counts correlation between bcbio-kallisto and dragen native output")

#Per gene counts correlation between dragen-txi and dragen-native outputs
cor.gene.dragentxi.dragennative = cor.test(log2(dragen.dragennative.gene.counts$dragen_txi+1),
                                 log2(dragen.dragennative.gene.counts$dragen_native+1), method = 'spearman')
ggplot(dragen.dragennative.gene.counts, 
       aes(x=log2(dragen.dragennative.gene.counts$dragen_txi+1), y=log2(dragen.dragennative.gene.counts$dragen_native+1))) +
         geom_point() +
         geom_smooth(method = "lm") +
         geom_abline(slope=1, intercept = 0, color="red") +
         annotate("text", x=15, y=20, label= paste("spearman cor = ", round(cor.gene.dragentxi.dragennative$estimate, 2))) +
         ggtitle("Per gene counts correlation between dragen-txi and dragen native output")

#summarize percentage difference between counts
transcripts.counts.summary <- summary(est.counts$percentage_counts_diff)
genes.count.summary <- summary(bcbio.dragen.gene.counts$percentageDiff)
genes.count.summary.dragennative <-  summary(bcbio.dragennative.gene.counts$percentageDiff)

#display results only where we have %diff above a certain threshold (using values above third quantile) for gene level counts
bcbio.dragen.gene.counts.filtered <- bcbio.dragen.gene.counts %>%
  dplyr::filter(percentageDiff > quantile(bcbio.dragen.gene.counts$percentageDiff, na.rm = TRUE)[3])

bcbio.dragennative.gene.counts.filtered <- bcbio.dragennative.gene.counts %>%
  dplyr::filter(percentageDiff > quantile(bcbio.dragennative.gene.counts$percentageDiff, na.rm = TRUE)[3])

dragen.dragennative.gene.counts.filtered <- dragen.dragennative.gene.counts %>%
  dplyr::filter(percentageDiff > quantile(dragen.dragennative.gene.counts$percentageDiff, na.rm = TRUE)[3])

datatable(bcbio.dragen.gene.counts.filtered, rownames = FALSE, filter="top", options = list(pageLength = 5, scrollX=T), caption = "Filtered gene counts (values above third quantile)")

datatable(bcbio.dragennative.gene.counts.filtered, rownames = FALSE, filter="top", options = list(pageLength = 5, scrollX=T), caption = "Filtered dragen native gene counts (values above third quantile)")

datatable(dragen.dragennative.gene.counts.filtered, rownames = FALSE, filter="top", options = list(pageLength = 5, scrollX=T), caption = "Filtered dragen txi & dragen native gene counts (values above third quantile)")

#generate interactive density plot
plotly::plot_ly(bcbio.dragen.gene.counts.filtered, x = ~log2(bcbio+1), y = ~log2(dragen+1), type = 'scatter', color = ~percentageDiff,
                mode= 'markers',
                text = ~paste('geneName:', geneName, 
                              '<br /> bcbioCounts: ', bcbio, 
                              '<br /> dragenCounts: ', dragen),
                hoverinfo = 'text') %>% 
  layout(title="Density plot for filtered counts <br> %age difference above third quantile")

plotly::plot_ly(bcbio.dragennative.gene.counts.filtered, x = ~log2(bcbio+1), y = ~log2(dragen_native+1), type = 'scatter', color = ~percentageDiff,
                mode= 'markers',
                text = ~paste('geneName:', geneName, 
                              '<br /> bcbioCounts: ', bcbio, 
                              '<br /> dragenNativeCounts: ', dragen_native),
                hoverinfo = 'text') %>% 
  layout(title="Density plot for filtered dragen native gene counts <br> %age difference above third quantile")

# subset the filtered counts to the genes in UMCCR cancer gene list
bcbio.dragen.gene.counts.filtered.umccr <- bcbio.dragen.gene.counts.filtered %>%
  dplyr::filter(geneName %in% umccr.cancer.genes$symbol)

datatable(bcbio.dragen.gene.counts.filtered.umccr, rownames = FALSE, filter="top", options = list(pageLength = 5, scrollX=T), caption = "Filtered gene counts that exist in UMCCR cancer gene list")

bcbio.dragennative.gene.counts.filtered.umccr <- bcbio.dragennative.gene.counts.filtered %>%
  dplyr::filter(geneName %in% umccr.cancer.genes$symbol)

datatable(bcbio.dragennative.gene.counts.filtered.umccr, rownames = FALSE, filter="top", options = list(pageLength = 5, scrollX=T), caption = "Filtered dragen native gene counts that exist in UMCCR cancer gene list")

dragen.dragennative.gene.counts.filtered.umccr <- dragen.dragennative.gene.counts.filtered %>%
  dplyr::filter(geneName %in% umccr.cancer.genes$symbol)

#generate interactive density plot

plotly::plot_ly(bcbio.dragen.gene.counts.filtered.umccr, x = ~log2(bcbio+1), y = ~log2(dragen+1), type = 'scatter', color = ~percentageDiff,
                mode= 'markers',
                text = ~paste('geneName:', geneName, 
                              '<br /> bcbioCounts: ', bcbio, 
                              '<br /> dragenCounts: ', dragen),
                hoverinfo = 'text') %>% 
  layout(title="Density plot for filtered counts <br>Genes present in cancer gene list")

plotly::plot_ly(bcbio.dragennative.gene.counts.filtered.umccr, x = ~log2(bcbio+1), y = ~log2(dragen_native+1), type = 'scatter', color = ~percentageDiff,
                mode= 'markers',
                text = ~paste('geneName:', geneName, 
                              '<br /> bcbioCounts: ', bcbio, 
                              '<br /> dragenNativeCounts: ', dragen_native),
                hoverinfo = 'text') %>% 
  layout(title="Density plot for filtered counts <br>Native dragen output genes present in cancer gene list")

plotly::plot_ly(dragen.dragennative.gene.counts.filtered.umccr, x = ~log2(dragen_txi+1), y = ~log2(dragen_native+1), type = 'scatter', color = ~percentageDiff,
                mode= 'markers',
                text = ~paste('geneName:', geneName, 
                              '<br /> dragentxiCounts: ', dragen_txi, 
                              '<br /> dragenNativeCounts: ', dragen_native),
                hoverinfo = 'text') %>% 
  layout(title="Density plot for filtered counts <br>Dragen txi and dragen native output genes present in cancer gene list")

```

### TPMs

Not showing scatter or violin plots for tpms atm - The initial analysis show the values are failrly comparable.
Also, in the report, we calculate TPMs from gene level count estimates. So the above data should suffice.

```{r tpm-plots, message=FALSE, warning=FALSE, eval = FALSE}
###merge both input dataframes on target_id but only keep tpm from second dataframe and target_id and tpms from first dataframe
icombined_tpm <- merge(x = counts.bcbio[,c('target_id', 'tpm')], y = counts.dragen[,c('target_id','tpm')], by = "target_id", all.x = TRUE)
colnames(icombined_tpm) <- c('target_id', 'tpm_bcbio', 'tpm_dragen')
###function calls
scatterPlot(icombined_tpm, main_title = "Trascripts per million (tpm) values across bcbio and dragen")
violinPlot(icombined_tpm, main_title = "Trascripts per million (tpm) values across bcbio and dragen runs")
```

***

# Fusions Analysis

This section includes a summary of:

- the total number of fusions in the compared files and 
- a table outlining difference between both runs.
  - The fields checked are fusion gene names.
  
```{r input-fusions, message=FALSE, warning=FALSE, eval=FALSE}
###compare the fusion calls (from both runs)
fusions.bcbio.arriba <- read.table(here("../../data/projects/WTS_comparison/SBJ02407/bcbio/fusions.tsv"), sep="\t", as.is=TRUE, header=FALSE, fill = TRUE)
colnames(fusions.bcbio.arriba) <- c("gene1", "gene2", "strand1(gene/fusion)", "strand2(gene/fusion)", "breakpoint1", "breakpoint", "site1", "site2", "type", "direction1", "direction2", "split_reads1", "split_reads2", "discordant_mates", "coverage1", "coverage2", "confidence", "closest_genomic_breakpoint1", "closest_genomic_breakpoint2", "filters", "fusion_transcript", "reading_frame", "peptide_sequence", "read_identifiers")
fusions.dragen.arriba <- read.table(here("../../data/projects/WTS_comparison/SBJ02407/dragen/fusions.tsv"), sep="\t", as.is=TRUE, header=FALSE, fill = TRUE)
colnames(fusions.dragen.arriba) <- c("gene1", "gene2", "strand1(gene/fusion)", "strand2(gene/fusion)", "breakpoint1", "breakpoint2", "site1", "site2", "type", "split_reads1", "split_reads2", "discordant_mates", "coverage1", "coverage2", "confidence", "reading_frame", "tags", "retained_protein_domains", "closest_genomic_breakpoint1", "closest_genomic_breakpoint2", "gene_id1", "gene_id2", "transcript_id1", "transcript_id2", "direction1", "direction2", "filters", "fusion_transcript", "peptide_sequence", "read_identifiers")
```


```{r diff-fusions, message=FALSE, warning=FALSE, eval=FALSE}
###find common fusion gene pairs between both arriba and dragen. Return all rows from x where there are matching values in y, and all columns from x and y. If there are multiple matches between x and y, all combination of the matches are returned.
fusions.common <- inner_join(fusions.bcbio.arriba, fusions.dragen.arriba, by = c("gene1", "gene2"), suffix = c(".bcbio", ".dragen"),)

###find different fusion gene pairs between both arriba and dragen. Return all rows from x where there are not matching values in y, keeping just columns from x.
fusions.diff.bcbio <- anti_join(fusions.bcbio.arriba, fusions.dragen.arriba, by = c("gene1", "gene2"))
fusions.diff.dragen <- anti_join(fusions.dragen.arriba, fusions.bcbio.arriba, by = c("gene1", "gene2"))
```

### Fusions Summary

```{r summary-fusions, message=FALSE, warning=FALSE, echo =FALSE, eval=FALSE}
fusions.total <- data.frame(c("bcbio-arriba", "dragen-arriba"), c(nrow(fusions.bcbio.arriba), nrow(fusions.dragen.arriba)))
colnames(fusions.total) <- c("pipeline", "total_fusions")
datatable(fusions.total, rownames = TRUE, options = list(sDom  = '<"top">lrt<"bottom">ip'))

fusions.summary <- data.frame(c("Number of common fusion calls", "Number of fusions only called by bcbio-arriba", "Number of calls only called by dragen-arriba"),
                              c(nrow(fusions.common), nrow(fusions.diff.bcbio), nrow(fusions.diff.dragen)))
colnames(fusions.summary) <- c("metrics", "value")
datatable(fusions.summary, rownames = TRUE, options = list(sDom  = '<"top">lrt<"bottom">ip'))

datatable(fusions.common, rownames = FALSE, filter="top", options = list(pageLength = 5, scrollX=T), caption = "Fusion gene pairs detected by both bcbio-arriba and dragen-arriba")
datatable(fusions.diff.bcbio, rownames = FALSE, filter="top", options = list(pageLength = 5, scrollX=T), caption = "Fusions only detected by bcbio-arriba")
datatable(fusions.diff.dragen, rownames = FALSE, filter="top", options = list(pageLength = 5, scrollX=T), caption = "Fusions only detected by dragen-arriba")

```


