---
title: "bcbio Native versus Cromwell results comparison"
author: "Sehrish Kanwal (UMCCR)"
date: "`r format(Sys.time(), '%a %Y-%b-%d')`"
output:
  html_document:
    css: style.css
    theme: readable
    toc: true
    toc_float: true
    code_folding: hide
  rmdformats::material:
    highlight: kate
params:
  run_1: '/Users/kanwals/Documents/UMCCR/data/WTS_comparison/X2016_249_17_MH_P033/bcbio_native/run_1/'
  run_2: '/Users/kanwals/Documents/UMCCR/data/WTS_comparison/X2016_249_17_MH_P033/bcbio_cromwell/run_1'
---

# Description

<style>
div.blue { background-color:#e6f0ff; border-radius: 5px; padding: 20px;}
</style>
<div class = "blue">

**SAMPLE: CCR170115b_MH17T002P033_RNA**

For each WTS workflow output type (e.g. counts, abundances, fusions), the document currently reports:

- File names in the input (parms) directories
- The number of calls
- The key difference between both outputs (e.g. bcbio native VS bcbio cromwell currently) in a tabular format

</div>


```{r setup, include=FALSE}
knitr::opts_chunk$set(collapse = TRUE) # global setup
```


```{r define_functions, comment=NA, message=FALSE, warning=FALSE}
### Define functions.
###list files in a directory
lf <- function(...) {
  data.frame(fname = list.files(...)) %>% 
    knitr::kable(row.names = TRUE)
}

###count number of lines in a file
countLines <- function(file_names) {
  target_file <- data.frame(matrix(ncol = length(counts_file_name), nrow = 0))

  for (i in 1:length(file_names)) {
    if (file_ext(file_names[i]) == 'counts') {
      file <- read.table(file_names[i], sep="\t", as.is=TRUE, header=FALSE)
    } else {
      file <- read.table(file_names[i], sep="\t", as.is=TRUE, header=TRUE)
    }
    #count lines
    target_file <- rbind(target_file, data.frame(basename(file_names[i]), nrow(file)))
  }
  colnames(target_file) <- c("file", "lines_count")
  rownames(target_file) <- c("bcbio_native", "bcbio_cromwell")
  return(target_file)
}

###return key differences (according to set criteria) between input files
diffFiles <- function(file_names) {
  #an anti join returns the rows of the first table where it cannot find a match in the second table.
  #also pass a character vector of variables to join by.
  if (stringr::str_detect(file_names[1], pattern = 'counts$')) {
    input1 <- read.table(file_names[1], sep="\t", as.is=TRUE, header=FALSE)
    colnames(input1) <- c('gene_id', 'count')
    input2 <- read.table(file_names[2], sep="\t", as.is=TRUE, header=FALSE)
    colnames(input2) <- c('gene_id', 'count')
    join_by <- c('gene_id', 'count')
  } else if (stringr::str_detect(file_names[1], pattern = 'flat.tsv$')){
    
    input1 <- read.table(file_names[1], sep="\t", as.is=TRUE, header=TRUE)
    input2 <- read.table(file_names[2], sep="\t", as.is=TRUE, header=TRUE)
    join_by <-  c('geneA.name', 'geneB.name', 'paircount', 'splitcount')
  } else if (stringr::str_detect(file_names[1], pattern = 'abundance.tsv$')){
    
    input1 <- read.table(file_names[1], sep="\t", as.is=TRUE, header=TRUE)
    input2 <- read.table(file_names[2], sep="\t", as.is=TRUE, header=TRUE)
    
    #merge both dataframes on target_id but only keep est_counts from second dataframe i.e. run2 (and all cols from first dataframe i.e. run1)
    icombined <- merge(x = input1, y = input2[,c(1,4)], by = "target_id", all.x = TRUE)
    icombined_filtered <- data.frame(matrix(ncol = 7, nrow = 0))
    colnames(icombined_filtered) <- c('target_id',	'length',	'eff_length',	'est_counts_run1',	'tpm', 'est_counts_run2', 'percent_change')
    #icomb_test <- data.frame(matrix(ncol = 3, nrow = 0))
    
    for (i in 1:nrow(icombined)) {
      #check for percentage decrease (to avoid negative differences) between estimated count values (https://www.skillsyouneed.com/num/percent-change.html)
      if(icombined[i,'est_counts.x'] != icombined[i,'est_counts.y']) {
        if (icombined[i,'est_counts.x'] > icombined[i,'est_counts.y']) {
        original_number <- icombined[i,'est_counts.x']
        new_number <- icombined[i,'est_counts.y']
      } else {
        original_number <- icombined[i,'est_counts.y']
        new_number <- icombined[i,'est_counts.x']
      percent_change <- ((original_number - new_number)/original_number) * 100
      #icomb_test <- rbind(icomb_test, c(original_number, new_number, percent_change))
      if ( percent_change > 5) {
        icombined_filtered <- rbind(icombined_filtered, icombined[i,])
        icombined_filtered[i, 'percent_change'] <- percent_change
      }
      }
      }
    colnames(icombined_filtered) <- c('target_id',	'length',	'eff_length',	'est_counts_run1',	'tpm', 'est_counts_run2', 'percent_change')
    icombined_filtered <- icombined_filtered[,c('target_id',	'length',	'eff_length',	'est_counts_run1',	'est_counts_run2', 'percent_change', 'tpm')]
    return(icombined_filtered)
    }
  }
  
  diff1 <- dplyr::anti_join(input1, input2, by = join_by) %>%
    mutate(input_info = 'bcbio_native') %>%
    dplyr::select(input_info, everything())
  
  diff2 <- dplyr::anti_join(input2, input1, by = join_by) %>%
    mutate(input_info = 'bcbio_cromwell') %>%
    dplyr::select(input_info, everything())
  
  return(rbind(diff1,diff2))
}

###generate scatter and violin plot for tpm values across both runs
scatterPlot <- function(icombined_tpm, main_title, plot_mode = "static") {
  plotly::plot_ly(icombined_tpm, x = ~log2(tpm_run1+0.5), y = ~log2(tpm_run2+0.5), type = 'scattergl', 
                       text = icombined_tpm$target_id, width = 00, height = 400,
                               marker = list(size = 10,
                       color = 'rgba(255, 182, 193, .5)',
                       #opacity = 0.5, https://plot.ly/~alex/455/four-ways-to-change-opacity-of-scatter-markers.embed
                       line = list(color = 'rgba(152, 0, 0, .8)',
                                   width = 2))) %>%
    plotly::layout(title = main_title)
  
  # Add Loess smoothed line - not a good idea for big data as this fitting makes it very slow
  #  plotly::add_lines(x = icombined_tpm$tpm_run1, y = ~fitted(loess(icombined_tpm$tpm_run1 ~ icombined_tpm$tpm_run2)),
  #              line = list(color = 'grey'),
  #              name = "Loess Smoother", opacity = 0.5, showlegend = TRUE) %>%
    
  #p_log <- ggplot(data = icombined_tpm, aes(x = log(tpm_run1), y = log(tpm_run2))) + 
    #geom_point(shape = 16, size = 2, show.legend = FALSE, color="firebrick", alpha = 0.1) 
  #p_log <- ggplotly(p_log) %>% toWebGL()
    
  #cowplot::plot_grid(p, p_log, labels = "AUTO")
}

###VIOLIN plot https://plot.ly/r/violin/
violinPlot <- function(icombined_tpm, main_title, plot_mode = "static") {
  #overall data distribution
  p1 <- plotly::plot_ly(icombined_tpm, type = 'violin') %>%
    add_trace(y = ~log2(tpm_run1+0.5), box = list(visible = T),  meanline = list(visible = F),
                       x = 'Sample_bcbio_native', legendgroup = 'run1', text = icombined_tpm$target_id, name = 'bcbio_native', color = 'orange') %>%
    add_trace(y = ~log2(tpm_run2+0.5), box = list(visible = T),  meanline = list(visible = F),
                       x = 'Sample_bcbio_cromwell', legendgroup = 'run2', text = icombined_tpm$target_id, name = 'bcbio_cromwell', color = 'green') %>%
    layout(yaxis = list(title = "log2 tpm", zeroline = F))
  
  #zoom into data to highlight data spread
  p2 <- plotly::plot_ly(icombined_tpm, type = 'violin') %>%
    add_trace(y = c(~log(quantile(tpm_run1)[1]+0.5), ~log(quantile(tpm_run1)[3]+0.5)+0.5), box = list(visible = T),  meanline = list(visible = F),
                       x = 'Sample_bcbio_native', legendgroup = 'run1', text = icombined_tpm$target_id, name = 'bcbio_native', color = 'orange') %>%
    
    add_trace(y = c(~log(quantile(tpm_run2)[1]+0.5), ~log(quantile(tpm_run2)[3]+0.5)+0.5), box = list(visible = T),  meanline = list(visible = F),
                       x = 'Sample_bcbio_cromwell', legendgroup = 'run2', text = icombined_tpm$target_id, name = 'bcbio_cromwell', color = 'green') %>%
    layout(yaxis = list(title = "log tpm", zeroline = F))
  
  #combine both box plots
  subplot(p1, p2, shareX = FALSE, shareY = FALSE, titleY = TRUE, margin = 0.06) %>%
    hide_legend()
}

```


```{r, message=FALSE, warning=FALSE}
### Required packages.
library(tools)
library(DT)
library(dplyr)
library(kableExtra)
library(ggplot2)
library(plotly)
library(gridExtra)
library(tximport)
library(tximportData)
library(package="EnsDb.Hsapiens.v75", character.only = TRUE)
```

***

# List files

This section lists the file names for outputs in the run directories being compared in this report.  

```{r read_data, message=FALSE, warning=FALSE}
###Read input. 
###The script expects two input directories, each containing 'final' results as produced by bcbio. The results in these two input directories will be compared.
run_1 <-file.path(params$run_1)
#list files in the first run directory
lf(run_1)

run_2 <- file.path(params$run_2)
#list files in the second run directory
lf(run_2)

```

***

# Counts

This section summarizes two key factors:

- a summary of the total number of counts in the compared files and 
- a table outlining difference between both runs.

```{r input-counts, message=FALSE, warning=FALSE}
###Start with the simple line count in the files i.e. the number of entries in the output from both runs is same.
#compare the count files first (from both runs)
counts_file_name <- dir(c(run_1, run_2), pattern = ".counts")
counts_file_name[1] <-  paste(run_1, counts_file_name[1], sep = "/")
counts_file_name[2] <-  paste(run_2, counts_file_name[2], sep = "/")
```

### Total counts

```{r total-counts, message=FALSE, warning=FALSE}
###functionc call for calculating number of lines in count files from both runs
lines_count <- countLines(counts_file_name)
datatable(lines_count, rownames = TRUE, options = list(sDom  = '<"top">lrt<"bottom">ip'))
```

### Diff counts

```{r diff-counts, message=FALSE, warning=FALSE}
###function call for finding the differences between count files from both runs
diff_count <- diffFiles(counts_file_name)
datatable(diff_count, rownames = FALSE, filter="top", options = list(pageLength = 5, scrollX=T))
```

***

# Abundance

This section summarizes following information:

- a summary of the total number of abundance entries in the compared files and 
- a table outlining difference between both runs. Percentage difference is calculated against the abundance measurements. 
  - The set threshold currenly is 5%. Any difference below this percent value is not flagged.
- Also, included are density plots for tpm (transcripts per million) field in the output, which is not used for calculating %age difference but is important to demonstrate overall data spread.
  - The first scatter plot compares log transformed tpm values across both runs. This would aid in providing "at a glance" overiew of data divergence (if any).
  - The violin plots is included as another representation of the same data. The first one is on the whole data (log tpm) values and the second one is focussing on probability density at first and third data quartiles (useful for understanding data range, specially in this case when the data spread is large).

```{r input-abundance, message=FALSE, warning=FALSE}
###compare the kallisto abundance calls (from both runs)
abundance_file_name <- dir(c(run_1, run_2), pattern = "abundance.tsv")
abundance_file_name[1] <-  paste(run_1, abundance_file_name[1], sep = "/")
abundance_file_name[2] <-  paste(run_2, abundance_file_name[2], sep = "/")
input1 <- read.table(abundance_file_name[1], sep="\t", as.is=TRUE, header=TRUE)
input2 <- read.table(abundance_file_name[2], sep="\t", as.is=TRUE, header=TRUE)
###merge both dataframes on target_id but only keep tpm from second dataframe i.e. run2 (and target_id and tpms from first dataframe i.e. run1)
icombined_tpm <- merge(x = input1[,c('target_id', 'tpm')], y = input2[,c('target_id','tpm')], by = "target_id", all.x = TRUE)
colnames(icombined_tpm) <- c('target_id', 'tpm_run1', 'tpm_run2')
###function calls
scatterPlot(icombined_tpm, main_title = "Trascripts per million (tpm) values across both runs")
violinPlot(icombined_tpm, main_title = "Trascripts per million (tpm) values across both runs")
```

### Total abundance

```{r total-abundance, message=FALSE, warning=FALSE}
###functionc call for calculating number of lines in fusion files form both runs
lines_abundance <- countLines(abundance_file_name)
datatable(lines_abundance, rownames = TRUE, options = list(sDom  = '<"top">lrt<"bottom">ip'))
```

### Diff abundance

```{r diff-abundance, message=FALSE, warning=FALSE}
###function call for finding the differences between fusion calls from both runs
diff_abundance <- diffFiles(abundance_file_name)
datatable(diff_abundance, rownames = FALSE, filter="top", options = list(pageLength = 5, scrollX=T))

#read files
input1 <- read.table(abundance_file_name[1], sep="\t", as.is=TRUE, header=TRUE) %>%
      mutate_at(4, round, 0)
input2 <- read.table(abundance_file_name[2], sep="\t", as.is=TRUE, header=TRUE) %>%
      mutate_at(4, round, 0)
#check if target_id column have same values in both dataframes
if (identical(input1[,1], input2[,1])) {
  
}
```

### Diff abundance salmon VS featurecounts
```{r diff2-abundance, message=FALSE, warning=FALSE}
####read salmon counts and featurecounts
salmon <- list.files("/Users/kanwals/Documents/UMCCR/data/WTS_comparison/X2016_249_17_MH_P033/bcbio_native/run_1/salmon/", pattern = "quant.sf", full.names = T)
featurecounts <- read.table("/Users/kanwals/Documents/UMCCR/data/WTS_comparison/X2016_249_17_MH_P033/bcbio_native/run_1/X2016_249_17_MH_P033-ready.counts", sep="\t", as.is=TRUE) 
colnames(featurecounts) <- c("gene", "count")

#read in the tximport data
dir <- system.file("extdata", package = "tximportData")
tx2ensembl <- read.csv(file.path(dir, "tx2gene.ensembl.v87.csv"))

#look at countsFromAbundance parameter to change the method to generate the counts
txi.salmon <- tximport(salmon, type = "salmon", tx2gene = tx2ensembl)

#extract salmon counts to prepare dataframe
salmon.counts <- as.data.frame(txi.salmon$counts) %>%
  tibble::rownames_to_column("gene") %>%
  dplyr::rename(count = V1)

#merge the calls from featurecounts and salmon - also add a column for recording percentage difference
feature.salmon.counts <- inner_join(featurecounts, salmon.counts, by = "gene") %>%
  dplyr::mutate(percentageDiff = ifelse(count.x > count.y, (count.x - count.y)/count.x *100, (count.y - count.x)/count.y *100)) %>%
  dplyr::mutate_if(is.numeric, round,2)
colnames(feature.salmon.counts) <- c("gene", "featurecounts", "salmoncounts", "percentageDiff") 

#get genes annotation from ensembl IDs 
edb <- EnsDb.Hsapiens.v75
#get keytypes for gene SYMBOL
keys <- keys(edb, keytype="GENEID")
#get genes genomic coordiantes
gene_info <- ensembldb::select(edb, keys=keys, columns=c("GENEID", "GENENAME"), keytype="GENEID")
colnames(gene_info) <- c("gene", "geneName")
#remove rows with duplicated ENSEMBL IDs - if any
gene_info = gene_info[!duplicated(gene_info$gene),]
gene_info = gene_info[!duplicated(gene_info$geneName),]

#annotate the counts dataframe to include genenames
counts <- inner_join(feature.salmon.counts, gene_info, by = "gene") %>%
  dplyr::arrange(-percentageDiff) %>%
  dplyr::select(geneName, everything()) %>%
  #remove part of genenames after .
  dplyr::mutate(geneName=sapply(strsplit(geneName, split = '.', fixed = TRUE), function(x) (x[1])))

#counts correlation between featurecounts and salmon counts 
ggplot(counts, 
       aes(x=log2(counts$featurecounts+1), y=log2(counts$salmoncounts+1))) +
         geom_point() +
         geom_smooth(method = "lm") +
         geom_abline(slope=1, intercept = 0, color="red") +
         annotate("text", x=15, y=20, label= "spearman cor = 0.95") +
         ggtitle("featurecounts versus salmon counts")

#summarize percentage difference between counts
counts.summary <- summary(counts$percentageDiff)

#Display results only where we have %diff above a certain threshold (using values above third qunartile)
counts.filtered <- counts %>%
  dplyr::filter(percentageDiff > quantile(counts$percentageDiff, na.rm = TRUE)[4])

datatable(counts.filtered, rownames = FALSE, filter="top", options = list(pageLength = 5, scrollX=T))

ggplot(counts.filtered, 
       aes(x=log2(counts.filtered$featurecounts+1), y=log2(counts.filtered$salmoncounts+1))) +
         geom_point() +
         geom_abline(slope=1, intercept = 0, color="red") +
         ggtitle("featurecounts versus salmon counts (filtered)")

# Generate interactive density plot
plotly::plot_ly(counts.filtered, x = ~log2(featurecounts+1), y = ~log2(salmoncounts+1), type = 'scatter', color = ~percentageDiff, mode= 'markers', 
                text = ~paste('geneName:', geneName, 
                              '<br /> featurecounts: ', featurecounts, 
                              '<br /> salmoncounts: ', salmoncounts),
                hoverinfo = 'text')

```

***

# Fusions

This section includes a summary of:

- the total number of fusions in the compared files and 
- a table outlining difference between both runs.
  - The fields checked are both fusion genes and pair/split count values (as this information is particularly critical for downstream postprocessing of WTS results).

```{r input-fusions, message=FALSE, warning=FALSE}
###compare the fusion calls (from both runs)
fusions_file_name <- dir(c(run_1, run_2), pattern = "flat.tsv")
fusions_file_name[1] <-  paste(run_1, fusions_file_name[1], sep = "/")
fusions_file_name[2] <-  paste(run_2, fusions_file_name[2], sep = "/")
```

### Total fusions

```{r total-fusions, message=FALSE, warning=FALSE}
###functionc call for calculating number of lines in fusion files form both runs
lines_fusion <- countLines(fusions_file_name)
datatable(lines_fusion, rownames = TRUE, options = list(sDom  = '<"top">lrt<"bottom">ip'))
```

### Diff fusions

```{r diff-fusions, message=FALSE, warning=FALSE}
###function call for finding the differences between fusion calls from both runs
diff_fusion <- diffFiles(fusions_file_name)
datatable(diff_fusion, rownames = FALSE, filter="top", options = list(pageLength = 5, scrollX=T))
```

***



