---
title: "WTS - bcbio BS dragen results"
author: "Sehrish Kanwal (UMCCR)"
date: "`r format(Sys.time(), '%a %Y-%b-%d')`"
output:
  html_document:
    css: style.css
    theme: readable
    toc: true
    toc_float: true
    code_folding: hide
  rmdformats::material:
    highlight: kate
params:
  run_1: '/Users/kanwals/Documents/UMCCR/data/WTS_comparison/MDX190024_UR3044456_T_rna/dragen/'
  run_2: '/Users/kanwals/Documents/UMCCR/data/WTS_comparison/MDX190024_UR3044456_T_rna/bcbio/'
---

# Description

<style>
div.blue { background-color:#e6f0ff; border-radius: 5px; padding: 20px;}
</style>
<div class = "blue">

**SAMPLE: MDX190024_UR3044456_T_rna**

For each WTS workflow output type (abundance and fusions for now), the document currently reports:

- File names in the input (parms) directories
- The number of calls
- The key difference between both outputs (i.e. bcbio VS dragen) in a tabular format

</div>


```{r setup, include=FALSE}
knitr::opts_chunk$set(collapse = TRUE) # global setup
```

```{r define_functions, comment=NA, message=FALSE, warning=FALSE}
### Define functions.
###list files in a directory
lf <- function(...) {
  data.frame(fname = list.files(...)) %>% 
    knitr::kable(row.names = TRUE)
}

```


```{r, message=FALSE, warning=FALSE}
### Required packages.
library(tools)
library(DT)
library(dplyr)
library(kableExtra)
library(ggplot2)
library(plotly)
library(gridExtra)
library(tximport)
library(package="EnsDb.Hsapiens.v75", character.only = TRUE)
```

***

# List files

This section lists the file names for outputs in the run directories being compared in this report.  

```{r read_data, message=FALSE, warning=FALSE}
###Read input. 
###The script expects two input directories, each containing 'final' results as produced by bcbio. The results in these two input directories will be compared.
run_1 <-file.path(params$run_1)
#list files in the first run directory
lf(run_1)

run_2 <- file.path(params$run_2)
#list files in the second run directory
lf(run_2)

```

***

# Counts

This section summarizes two key factors:

- a summary of the total number of counts in the compared files and 
- a table outlining difference between both runs.

```{r input-counts, message=FALSE, warning=FALSE}
###Start with the simple line count in the files i.e. the number of entries in the output from both runs is same.
#compare the count files first (from both runs)
counts.dragen <- read.table("/Users/kanwals/Documents/UMCCR/data/WTS_comparison/MDX190024_UR3044456_T_rna/dragen/UR3044456_WTS.quant.sf", sep="\t", as.is=TRUE, header=TRUE)
counts.bcbio <-  read.table("/Users/kanwals/Documents/UMCCR/data/WTS_comparison/MDX190024_UR3044456_T_rna/bcbio/abundance.tsv", sep="\t", as.is=TRUE, header=TRUE)

#read in the tximport data
dir <- "/Users/kanwals/Documents/UMCCR/data/WTS_comparison"
tx2ensembl <- read.csv(file.path(dir, "tx2gene.ensembl.v95.csv"), header = FALSE)
colnames(tx2ensembl) <- c("tx_name", "gene_id")
```

### Total counts

```{r total-counts, message=FALSE, warning=FALSE}
###functionc call for calculating number of lines in count files from both runs
lines.count<- data.frame(c("dragen", "bcbio"), c(nrow(counts.dragen), nrow(counts.bcbio)))
colnames(lines.count) <- c("file", "lines_count")
datatable(lines.count, rownames = TRUE, options = list(sDom  = '<"top">lrt<"bottom">ip'))
```

### Diff counts

```{r diff-counts, message=FALSE, warning=FALSE}
###find the differences between count files from both runs
#dragen names columns differently. Changing these to reflect bcbio's output format, making the comaprison's easy.
colnames(counts.dragen) <-  c("target_id", "length", "eff_length", "tpm", "est_counts")
counts.dragen <- counts.dragen[, c(1, 2, 3, 5, 4)]

#extract gene level counts using tximport
txi.bcbio <- tximport('/Users/kanwals/Documents/UMCCR/data/WTS_comparison/MDX190024_UR3044456_T_rna/bcbio/abundance.tsv', type = "kallisto", tx2gene = tx2ensembl)
txi.dragen <- tximport('/Users/kanwals/Documents/UMCCR/data/WTS_comparison/MDX190024_UR3044456_T_rna/dragen/UR3044456_WTS.quant.sf', type = "salmon", tx2gene = tx2ensembl)

#extract counts to prepare dataframe
bcbio.counts <- as.data.frame(txi.bcbio$counts) %>%
  tibble::rownames_to_column("gene") %>%
  dplyr::rename(count = V1)

dragen.counts <- as.data.frame(txi.dragen$counts) %>%
  tibble::rownames_to_column("gene") %>%
  dplyr::rename(count = V1)

#merge the calls from bcbio and dragen - also add a column for recording percentage difference
bcbio.dragen.counts <- inner_join(bcbio.counts, dragen.counts, by = "gene") %>%
  dplyr::mutate(percentageDiff = ifelse(count.x > count.y, (count.x - count.y)/count.x *100, (count.y - count.x)/count.y *100)) %>%
  dplyr::mutate_if(is.numeric, round,2)
colnames(bcbio.dragen.counts) <- c("gene", "bcbio", "dragen", "percentageDiff") 

#get genes annotation from ensembl IDs 
edb <- EnsDb.Hsapiens.v75
#get keytypes for gene SYMBOL
keys <- keys(edb, keytype="GENEID")
#get genes genomic coordiantes
gene_info <- ensembldb::select(edb, keys=keys, columns=c("GENEID", "GENENAME"), keytype="GENEID")
colnames(gene_info) <- c("gene", "geneName")
#remove rows with duplicated ENSEMBL IDs - if any
gene_info = gene_info[!duplicated(gene_info$gene),]
gene_info = gene_info[!duplicated(gene_info$geneName),]

#annotate the counts dataframe to include genenames
bcbio.dragen.counts <- inner_join(bcbio.dragen.counts, gene_info, by = "gene") %>%
  dplyr::arrange(-percentageDiff) %>%
  dplyr::select(geneName, everything()) %>%
  #remove part of genenames after .
  dplyr::mutate(geneName=sapply(strsplit(geneName, split = '.', fixed = TRUE), function(x) (x[1])))

##merge the raw estimated calls from bcbio and dragen - also add a column for recording percentage difference
dragen.bcbio.counts <- inner_join(counts.bcbio, counts.dragen, by = "target_id") %>%
  dplyr::mutate(percentageDiff = ifelse(est_counts.x > est_counts.y, (est_counts.x - est_counts.y)/est_counts.x *100, (est_counts.y - est_counts.x)/est_counts.y *100)) %>%
  dplyr::mutate_if(is.numeric, round,2)
colnames(dragen.bcbio.counts) <- c("target_id", "length.bcbio", "eff_length.bcbio", "est_counts.bcbio", "tpm.bcbio", "length.dragen", "eff_length.dargen", "est_counts.dragen", "tpm.dragen", "percentage_counts_diff")
#subset the df to only count values and percentage difference
est.counts <- dragen.bcbio.counts[, c('target_id', 'est_counts.bcbio', 'est_counts.dragen', 'percentage_counts_diff')]

#Per transcript counts correlation between bcbio-kallisto and dragen 
ggplot(counts, 
       aes(x=log2(est.counts$est_counts.bcbio+1), y=log2(est.counts$est_counts.dragen+1))) +
         geom_point() +
         geom_smooth(method = "lm") +
         geom_abline(slope=1, intercept = 0, color="red") +
         annotate("text", x=15, y=20, label= "spearman cor = 0.79") +
         ggtitle("featurecounts versus salmon counts")

#Per gene counts correlation between bcbio-kallisto and dragen 
ggplot(bcbio.dragen.counts, 
       aes(x=log2(bcbio.dragen.counts$bcbio+1), y=log2(bcbio.dragen.counts$dragen+1))) +
         geom_point() +
         geom_smooth(method = "lm") +
         geom_abline(slope=1, intercept = 0, color="red") +
         annotate("text", x=15, y=20, label= "spearman cor = 0.99") +
         ggtitle("Counts correlation between bcbio-kallisto and dragen")

#summarize percentage difference between counts
transcripts.counts.summary <- summary(est.counts$percentage_counts_diff)
genes.count.summary <- summary(bcbio.dragen.counts$percentageDiff)



```

***



